viewing_party/party.py

```diff
@@ -1,23 +1,144 @@
+        for movie in friend["watched"]:
+            friends_watched.add(movie["title"])
+    for title in friends_watched:
+        if title in user_watched:
+            user_watched.pop(title)
+    return list(user_watched.values())
+
+def get_friends_unique_watched(user_data):
+    """Return all the movies that at least one of the user's friends has 
+    watched but the user has not watched."""
+    friends_watched = {}
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_watched[movie["title"]] = movie
+    user_watched = {movie["title"] for movie in user_data["watched"]}
+    for title in user_watched:
+        if title in friends_watched:
+            friends_watched.pop(title)
+    return list(friends_watched.values())
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+    """Return a list of movies that the user has not watched, that one of 
+    their friends has watched, and that is available on a service they are 
+    subscribed to."""
+    # Build list of movies that friends have watched.
+    friend_movies = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie not in friend_movies:
+                friend_movies.append(movie)
+    # Filter out movies not on a service the user is subscribed to.
+    recommended_movies = []
+    subscriptions = user_data["subscriptions"]
+    watched = user_data["watched"]
+    for movie in friend_movies:
+        if movie not in watched and movie["host"] in subscriptions:
+            recommended_movies.append(movie)
+    return recommended_movies
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def get_new_rec_by_genre(user_data):
+    """Return a list of recommended movies in the user's most watched genre."""
+    if not user_data["watched"]:
+        return []
+    most_watched = get_most_watched_genre(user_data)
+    available_recs = get_available_recs(user_data)
+    new_recs = filter(
+        lambda movie: movie["genre"] == most_watched, 
+        available_recs
+    )
+    return list(new_recs)
+
+def get_rec_from_favorites(user_data):
+    """Return a list of movies that are in the user's favorites and that 
+    haven't been watched by their friends."""
+    # Get set of all movies that have been watched by friends.
+    friends_watched = set()
+    for friend in user_data["friends"]:
+        titles = [movie["title"] for movie in friend["watched"]]
+        friends_watched = friends_watched.union(titles)
+    # Get list of user's favorite movies that have not been watched by their 
+    # friends.
+    recs = filter(
+        lambda movie: movie["title"] not in friends_watched,
+        user_data["favorites"]
+    )
+    return list(recs)

```

Comment:
👍  This solution is so clean! Great work Eve!

Reply:
How does creating the `set` here affect the time and space complexity for this function?